snapcraft/internal/build_providers/_qemu/_qemu_driver.py

```diff
@@ -0,0 +1,175 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import shlex
+import shutil
+import socket
+import subprocess
+import sys
+import telnetlib
+import time
+from typing import List, Sequence
+
+import paramiko
+
+from snapcraft.internal.build_providers import errors
+
+
+logger = logging.getLogger(__name__)
+# Avoid getting paramiko logs which are overly verbose.
+logging.getLogger('paramiko').setLevel(logging.CRITICAL)
+
+
+def _popen(command: List) -> subprocess.Popen:
+    logger.debug('Running {}'.format(' '.join(command)))
+    return subprocess.Popen(command)
+
+
+def _get_qemu_command() -> str:
+    # TODO support more architectures.
+    return 'qemu-system-x86_64'
+
+
+class QemuDriver:
+    """A driver to interact with qemu virtual machines."""
+
+    provider_name = 'qemu'
+    _PROJECT_DEV = 'project_dev'
+    _PROJECT_MOUNT = 'project_mount'
+
+    def __init__(self, *, ssh_username: str, ssh_key_file: str) -> None:
+        """Initialize a QemuCommand instance.
+
+        :raises errors.ProviderCommandNotFound:
+            if the relevant qemu command is not found.
+        """
+        provider_cmd = _get_qemu_command()
+        if not shutil.which(provider_cmd):
+            raise errors.ProviderCommandNotFound(command=provider_cmd)
+        self.provider_cmd = provider_cmd
+        # TODO detect collisions and make dynamic
+        self._telnet_port = 64444
+        # TODO detect collisions and make dynamic
+        self._ssh_port = 5555
+        self._ssh_username = ssh_username
+        self._ssh_key_file = ssh_key_file
+        self._qemu_proc = None  # type: subprocess.Popen
+        self._ssh_handle = paramiko.SSHClient()
+        self._ssh_handle.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+    def launch(self, *, hda: str, qcow2_drives: Sequence,
+               project_9p_dev: str, ram: str=None,
+               enable_kvm: bool=True) -> None:
+        # TODO check for latest snapshot to launch for it instead of a cold
+        #      boot.
+        # TODO replace -nographic with a spinner and redirect to a log file.
+        cmd = [
+            'sudo', self.provider_cmd,
+            '-m', ram, '-nographic',
+            '-monitor', 'telnet::{},server,nowait'.format(self._telnet_port),
```

Comment:
I'm curious why we're using telnet here as opposed to stdio. Does it make things easier, even though we need to deal with the potential port conflicts?

Reply:
How do we talk to the backplane over stdio without mixing up common queries and output we do with snapcraft?